File: src/edu/neu/musicId/util/Utilities.java

package edu.neu.musicId.util;

/**
 * This class contains various static utility methods.
 * 
 * @author ckolek
 * @since 1.0
 */
public final class Utilities {
    /** Size of 16-bit unsigned integer in bytes */
    public static final int UINT16_SIZE = Short.SIZE / Byte.SIZE;

    /** Size of 32-bit unsigned integer in bytes */
    public static final int UINT32_SIZE = Integer.SIZE / Byte.SIZE;

    /** Size of float in bytes */
    public static final int FLOAT_SIZE = Float.SIZE / Byte.SIZE;

    private Utilities() {

    }

    /**
     * Extracts a 16-bit unsigned integer from the given array of
     * <code>byte</code>s offset at the given index.
     * 
     * @param bytes the array of <code>byte</code> values
     * @param offset the offset from the beginning of the array to extract the
     *            value from
     * @param isLittleEndian if the value is stored in little-endian form
     * @return an <code>int</code> with the 16-bit unsigned integer value
     * @since 1.0
     */
    public static int toInt16(byte[] bytes, int offset, boolean unsigned, boolean isLittleEndian) {
        return (int) toInt(bytes, offset, UINT16_SIZE, unsigned, isLittleEndian);
    }

    /**
     * Extracts a 16-bit unsigned integer from the given array of
     * <code>byte</code>s.
     * 
     * @param bytes the array of <code>byte</code> values
     * @param isLittleEndian if the value is stored in little-endian form
     * @return an <code>int</code> with the 16-bit unsigned integer value
     * @since 1.0
     */
    public static int toInt16(byte[] bytes, boolean unsigned, boolean isLittleEndian) {
        return toInt16(bytes, 0, unsigned, isLittleEndian);
    }

    /**
     * Extracts a 32-bit unsigned integer from the given array of
     * <code>byte</code>s offset at the given index.
     * 
     * @param bytes the array of <code>byte</code> values
     * @param offset the offset from the beginning of the array to extract the
     *            value from
     * @param isLittleEndian if the value is stored in little-endian form
     * @return a <code>long</code> with the 32-bit unsigned integer value
     * @since 1.0
     */
    public static long toInt32(byte[] bytes, int offset, boolean unsigned, boolean isLittleEndian) {
        return toInt(bytes, offset, UINT32_SIZE, unsigned, isLittleEndian);
    }

    /**
     * Extracts a 32-bit unsigned integer from the given array of
     * <code>byte</code>s.
     * 
     * @param bytes the array of <code>byte</code> values
     * @param isLittleEndian if the value is stored in little-endian form
     * @return a <code>long</code> with the 32-bit unsigned integer value
     * @since 1.0
     */
    public static long toInt32(byte[] bytes, boolean unsigned, boolean isLittleEndian) {
        return toInt32(bytes, 0, unsigned, isLittleEndian);
    }

    /**
     * Extracts a <code>length</code>-byte integer from the given array of
     * <code>byte</code>s offset at the given index.
     * 
     * @param bytes the array of <code>byte</code> values
     * @param offset
     * @param length
     * @param unsigned if the resulting value is unsigned
     * @param isLittleEndian if the value is stored in little-endian form
     * @return a <code>long</code> with the integer value
     * @since 1.0
     */
    public static long toInt(byte[] bytes, int offset, int length, boolean unsigned,
            boolean isLittleEndian) {
        long value = 0;

        for (int i = 0; i < length; i++) {
            byte b = isLittleEndian ? bytes[offset + i] : bytes[offset + (length - 1 - i)];

            if (unsigned) {
                value += (0xFF & b) << (Byte.SIZE * i);
            } else {
                value += b << (Byte.SIZE * i);
            }
        }

        return value;
    }
}

File: util/utilities.py
__author__ = 'ckolek'

import struct


def unpack_from(data, offset, is_unsigned, signed_format, unsigned_format,
                is_little_endian):
    """
    Extracts a simple value from an array of bytes.

    :param data: the array of bytes to extract the value from
    :param offset: the offset (in bytes) from the start of the array
    :param is_unsigned: True if the extracted value should be unsigned, False
     otherwise
    :param signed_format: The format string used in struct.unpack_from if the
     extracted value is signed
    :param unsigned_format: The format string used in struct.unpack_from if the
     extracted value is unsigned
    :param is_little_endian: True if the value is encoded in little-endian
     format, False otherwise
    :return: the extracted value
    """

    return struct.unpack_from(
        ('<' if is_little_endian else '>') +
        (unsigned_format if is_unsigned else signed_format),
        data,
        offset)[0]


def to_short(data, offset=0, is_unsigned=False, is_little_endian=True):
    """
    Extracts a short (16-bit) integer from an array of bytes.

    :param data: the array of bytes to extract the value from
    :param offset: the offset (in bytes) from the start of the array
    :param is_unsigned: True if the extracted value should be unsigned, False
     otherwise
    :param is_little_endian: True if the value is encoded in little-endian
     format, False otherwise
    :return: the extracted short value
    """

    return unpack_from(data, offset, is_unsigned, 'h', 'H', is_little_endian)


def to_int(data, offset=0, is_unsigned=False, is_little_endian=True):
    """
    Extracts an integer from an array of bytes.

    :param data: the array of bytes to extract the value from
    :param offset: the offset (in bytes) from the start of the array
    :param is_unsigned: True if the extracted value should be unsigned, False
     otherwise
    :param is_little_endian: True if the value is encoded in little-endian
     format, False otherwise
    :return: the extracted value
    """
    return unpack_from(data, offset, is_unsigned, 'i', 'I', is_little_endian)


def to_long(data, offset=0, is_unsigned=False, is_little_endian=True):
    """
    Extracts a long (32-bit) integer from an array of bytes.

    :param data: the array of bytes to extract the value from
    :param offset: the offset (in bytes) from the start of the array
    :param is_unsigned: True if the extracted value should be unsigned, False
     otherwise
    :param is_little_endian: True if the value is encoded in little-endian
     format, False otherwise
    :return: the extracted long value
    """
    return unpack_from(data, offset, is_unsigned, 'l', 'L', is_little_endian)


def to_integer(data, length, offset=0, is_unsigned=False,
               is_little_endian=True):
    """
    Extracts an integer value of a given length (in bytes) from the given array
     of bytes.

    :param data: the array of bytes to extract the value from
    :param length: the length (in bytes) of the value
    :param offset: the offset (in bytes) from the start of the array
    :param is_unsigned: True if the extracted value should be unsigned, False
     otherwise
    :param is_little_endian: True if the value is encoded in little-endian
     format, False otherwise
    :return: the extracted value
    """

    value = 0

    for i in xrange(length):
        b = ord(data[offset + i] if is_little_endian else data[
            offset + length - 1 - i])

        if is_unsigned:
            value += (0xFF & b) << (8 * i)
        else:
            value += b << (8 * i)

    return value